subsys/debug/cpu_load/cpu_load.c

```diff
@@ -0,0 +1,234 @@
+			      CONFIG_CPU_LOAD_LOG_INTERVAL);
+}
+
+void timer_handler(nrf_timer_event_t event_type, void *context)
+{
+	/*empty*/
+}
+
+int cpu_load_init(void)
+{
+	u8_t ch_sleep;
+	u8_t ch_wakeup;
+	u8_t ch_tick = 0;
+	nrfx_err_t err;
+	nrfx_timer_config_t config = NRFX_TIMER_DEFAULT_CONFIG;
+	int ret = 0;
+
+	config.frequency = NRF_TIMER_FREQ_1MHz;
+	config.bit_width = NRF_TIMER_BIT_WIDTH_32;
+
+	if (IS_ENABLED(CONFIG_CPU_LOAD_ALIGNED_CLOCKS)) {
+		/* It's assumed that RTC1 is driving system clock. */
+		config.mode = NRF_TIMER_MODE_COUNTER;
+		err = ppi_alloc(&ch_tick,
+		       nrf_rtc_event_address_get(NRF_RTC1, NRF_RTC_EVENT_TICK));
+		if (err != NRFX_SUCCESS) {
+			return -ENODEV;
+		}
+		nrfx_gppi_channel_endpoints_setup(ch_tick,
+		     nrf_rtc_event_address_get(NRF_RTC1, NRF_RTC_EVENT_TICK),
+		     nrfx_timer_task_address_get(&timer, NRF_TIMER_TASK_COUNT));
+		nrf_rtc_event_enable(NRF_RTC1, NRF_RTC_INT_TICK_MASK);
+	}
+
+	if (ready) {
+		return 0;
+	}
+
+	err = nrfx_timer_init(&timer, &config, timer_handler);
+	if (err != NRFX_SUCCESS) {
+		return -EBUSY;
+	}
+
+	err = ppi_alloc(&ch_sleep,
+		       nrf_power_event_address_get(NRF_POWER,
+						   NRF_POWER_EVENT_SLEEPENTER));
+	if (err != NRFX_SUCCESS) {
+		return -ENODEV;
+	}
+
+	err = ppi_alloc(&ch_wakeup,
+		       nrf_power_event_address_get(NRF_POWER,
+						   NRF_POWER_EVENT_SLEEPEXIT));
+	if (err != NRFX_SUCCESS) {
+		/* ppi free */
+		return -ENODEV;
+	}
+
+	nrfx_gppi_channel_endpoints_setup(ch_sleep,
+		  nrf_power_event_address_get(NRF_POWER,
+					      NRF_POWER_EVENT_SLEEPENTER),
+		  nrfx_timer_task_address_get(&timer, NRF_TIMER_TASK_START));
+	nrfx_gppi_channel_endpoints_setup(ch_wakeup,
+		  nrf_power_event_address_get(NRF_POWER,
+					      NRF_POWER_EVENT_SLEEPEXIT),
+		  nrfx_timer_task_address_get(&timer, NRF_TIMER_TASK_STOP));
+
+	nrfx_gppi_channels_enable((1 << ch_sleep) |
+				  (1 << ch_wakeup) |
+				  (1 << ch_tick));
+
+	cpu_load_reset();
+
+	if (IS_ENABLED(CONFIG_CPU_LOAD_LOG_PERIODIC)) {
+		ret = cpu_load_log_init();
+	}
+
+	ready = true;
+
+	return ret;
+}
+
+static u32_t cycle_ref;
+
+void cpu_load_reset(void)
+{
+	nrfx_timer_enable(&timer);
+	nrfx_timer_disable(&timer);
+	cycle_ref = k_cycle_get_32();
+}
+
+static u32_t sleep_ticks_to_us(u32_t ticks)
+{
+	return IS_ENABLED(CONFIG_CPU_LOAD_ALIGNED_CLOCKS) ?
+	   (u32_t)(((u64_t)ticks * 1000000) / CONFIG_SYS_CLOCK_TICKS_PER_SEC) :
+	   ticks;
+}
+
+u32_t cpu_load_get(void)
+{
+	u32_t sleep_us = sleep_ticks_to_us(nrfx_timer_capture(&timer, 0));
+	u32_t total_cyc = k_cycle_get_32() - cycle_ref;
+	u64_t total_us = ((u64_t)total_cyc * 1000000)/
+				CONFIG_SYS_CLOCK_TICKS_PER_SEC;
+	u64_t load;
+
+	__ASSERT(total_us < UINT32_MAX, "Measurement is limited.");
+
+	/* Because of different clock sources for system clock and TIMER it
+	 * is possible that sleep time is bigger than total measured time.
+	 *
+	 * Note that for simplicity reasons module does not handle TIMER
+	 * overflow.
```

Comment:
Perhaps it could be mentioned in the API description that a 32-bit register is used for counting microseconds of sleep time and thus `cpu_load_reset()` should be called from time to time (at least once per ~4294 seconds of sleeping)?

Reply:
added.